### PR TITLE
Add RAGAS evaluation metrics

### LINKED ISSUE #317: Add missing metrics and align existing from Ragas library
implement / fix current metric inside the repo without Ragas dependency

### PR DESCRIPTION
**Pull request by kcortinas**
closes #317 

This PR integrates [RAGAS](https://docs.ragas.io/en/latest/index.html) open source evaluation framework for RAG pipelines. The new added metrics offer support for both end-to-end and component-wise RAG evaluation:
- **End-to-end evaluation:** answer correctness, answer semantic similarity.
- **Component-wise evaluation:** retrieval metrics (context precision, context relevancy, context recall, context entity recall), generation metrics (faithfulness, answer relevancy).

This PR also fixes:
- Experiment name input when calling `generate_metrics`
- Missing params when calling `query_and_eval_single_line`

image

image

### CODE REVIEW

**Review by martinpeck — COMMENTED** [review]
This looks good to me.

I would rather @ritesh-modi give you a final approval, but from my point of view this should be approved and merged.

File: .github/workflows/config.json
Comment (prvenk) [inline .github/workflows/config.json:49]: Hey @kcortinas Should these be removed? maybe Im missing something but is the goal to leverage ragas version of answer and context relevance as opposed to the existing implementation or have both versions to pick from? because it looks like `llm_answer_relevance` and `llm_context_relevance` is still being referenced in some scripts like `eval.py`

Comment (martinpeck) [reply .github/workflows/config.json:49]: Not exactly answering your question, but they're being removed from this particular config file (the config used for CI automation builds), but they can still be used by end users.

The sample config.json retains these items.

@kcortinas : is the idea to only test with these new metrics, and not with the older ones? Or, should we run CI tests with *all* metrics?

Comment (prvenk) [reply .github/workflows/config.json:49]: ah got it.

**Comment by martinpeck** [comment]
What was the reason for the force-push?

File: .github/workflows/config.json
Comment (kcortinas) [reply .github/workflows/config.json:49]: @prvenk This PR focuses on adding the RAGAS integration. Despite the name, some llm-based metrics are not the same as the corresponding RAGAS ones, another PR will follow to adjust and/or remove redundant metrics.

@martinpeck  Good point! Just kept the old metrics in the CI config but had to adjust the logging as we were hitting the max length for the “run_metrics” parameter (no need to log the “_run_metrics_” dict as we’re already logging the metrics). I haven’t added the metrics `"ragas_answer_relevancy"`, `"ragas_answer_correctness"` and `"ragas_answer_similarity"` yet as they require the deployment of the embedding model `“text-embedding-ada-002”` in the subscription we’re using to run the CI pipeline: @ritesh-modi  is this a model we can deploy? Thanks!

**Comment by kcortinas** [comment]
> What was the reason for the force-push?

@martinpeck The branch became outdated so rebased and pushed with --force-with-lease to avoid overwriting

File: .github/workflows/config.json
Comment (martinpeck) [reply .github/workflows/config.json:49]: I'm 99.9% sure it's already deployed, and being used for other things, so I think you can assume that we have this deployed already.

Comment (guybartal) [reply .github/workflows/config.json:49]: I've also noticed that we were logging the `run_metrics` as dict and removed it in my [PR](https://github.com/microsoft/rag-experiment-accelerator/pull/529),
is that needed to be logged twice? @ritesh-modi

Comment (guybartal) [reply .github/workflows/config.json:49]: I know customers might already use the old metrics,
quick question, but are those old metrics important to keep?
aren't ragas metrics enough?
keeping code means maintain it, along with unit tests and dependencies...
we should consider removing all of them if we are sure that RAGAS is enough.

Comment (ritesh-modi) [reply .github/workflows/config.json:49]: We can leave 2-3 metrics (ragas an non-rags) as an example in config.json. It need not be only ragas

File: rag_experiment_accelerator/evaluation/eval.py
Comment (kcortinas) [reply rag_experiment_accelerator/evaluation/eval.py:429]: No. Since RAGAS can be used with different LLM models these wrappers help the RAGAS library implement a common interface used by the metrics. We could of course create our own implementation of the wrappers, but it would require duplicating a lot of existing logic and continuous monitoring of the RAGAS library for compatibility issues. Is there a specific reason why we would want not to import them?

File: .github/workflows/config.json
Comment (kcortinas) [reply .github/workflows/config.json:50]: Updated target branch to prerelease

**Review by guybartal — APPROVED** [review]
LGTM, if all comments are resolved, lets merge this to prerelease branch

File: rag_experiment_accelerator/evaluation/eval.py
Comment (ritesh-modi) [reply rag_experiment_accelerator/evaluation/eval.py:429]: This needs to be evaluated because of new security guidance

Comment (guybartal) [reply rag_experiment_accelerator/evaluation/eval.py:429]: @ritesh-modi , we have many LangChain dependencies already which we need to resolve with priority,
I think we should open a new separate issue to address this across and not block this PR to prerelease branch.
WDYT?

**Comment by kcortinas** [comment]
As discussed, closing this PR per security guidance

**Pull request closed without merging**
